### PR TITLE
Submodule gs unstructured splines

### PR DESCRIPTION
**Pull request by hverhelst**
Adds submodules and updates gsMappedSplines/gsMappedBasis. 
Compatible with the (to be made public) gsUnstructuredSplines module

-----

Pull requests can only be merged after at least one review (and
approval) from @gismo/admins.

Code submitted to the stable branch should be clean, well-documented
and free of bugs.

# Please consider the following checklist before issuing a pull request:
- [ ] Have you added an explanation of what your changes do and why
  you'd like us to include them?
- [ ] Have you documented any new codes using Doxygen comments?
- [ ] Have you written new tests or examples for your changes?
-----

### CODE REVIEW

**Review by hverhelst — COMMENTED** [review]
There are some new files in gsUniversal; how did they come up in this branch?
@mmoelle1, any idea?

Furhtermore, the pMultiGrid_example.cpp and pMultiGrid_example.dox get removed, is there a reason? Seems that in [this](https://github.com/gismo/gismo/commit/7810cca646d53dd2a214686ab25c1c029424f121) commit some things were removed, @weinmueller  do you know more? I had similar issues with kirchhoff-Love_example which I had to revert manually right now.

**Comment by weinmueller** [comment]
The files, pMultiGrid_example.cpp and pMultiGrid_example.dox, are added later to stable. I guess if we do the PR #510 into stable, then the pMultiGrid_example.cpp will be added. So I hope there should be no problem with pMultiGrid_example.cpp and pMultiGrid_example.dox. Those files were deleted when I merged stable into submodules and reverted this again (=remove these files).

**Comment by hverhelst** [comment]
Okay good, @weinmueller . 
Tests are now failing because the geometry `simple.xml` is removed. Is there a reason for that?

**Comment by weinmueller** [comment]
No there is no reason. Can you add this file? It is needed in `geometry_example.cpp`.

**Comment by weinmueller** [comment]
When do we finish this PR? And also with the submodules? Otherwise stable get more and more diverged.

File: examples/gsView.cpp [hunk 1]
```diff
@@ -19,7 +19,7 @@ using namespace gismo;
```
Comment (hverhelst) [inline examples/gsView.cpp]: Revert the changes in this file

File: README.md [hunk 2]
```diff
@@ -9,6 +9,7 @@
 
 ======================================================================
 =====             Geometry plus Simulation modules               =====
+=====                     version 21.12 Alpha                    =====
```
Comment (hverhelst) [inline README.md]: Remove this line?

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,6 @@
 # Submodule files
 submodules.txt
+extensions/
```
Comment (hverhelst) [inline .gitignore]: Revert this file

File: deploy/README.md [hunk 4]
```diff
@@ -1,23 +1,25 @@
+======================================================================
```
Comment (hverhelst) [inline deploy/README.md]: Revert file

File: doc/gismo.bib [hunk 5]
```diff
@@ -204,39 +204,3 @@ @article{Shahbazi2005
 author = "Khosro Shahbazi"
 }
```
Comment (hverhelst) [inline doc/gismo.bib]: Revert file

File: doc/Examples.dox [hunk 6]
```diff
@@ -54,8 +54,6 @@ In the gismo/examples sub-directory we find:
```
Comment (hverhelst) [inline doc/Examples.dox]: Revert file

File: doc/pMultiGrid_example.dox [hunk 7]
```diff
@@ -1,156 +0,0 @@
-namespace gismo {
```
Comment (hverhelst) [inline doc/pMultiGrid_example.dox]: Revert file

File: examples/biharmonic2_example.cpp [hunk 8]
```diff
@@ -107,7 +107,7 @@ int main(int argc, char *argv[])
     //! [Parse command line]
```
Comment (hverhelst) [inline examples/biharmonic2_example.cpp]: Revert file

**Review by hverhelst — COMMENTED** [review]
Changes in this commit:
1. Changes to gsFunctionSet for boundary conditions. Affects:
- `src/gsPde/gsBoundaryConditions.h`
- `src/gsPde/gsBoundaryConditions.hpp`
- `src/gsPde/gsBoundaryConditions_.cpp`
- `unittests/gsBoundaryConditions_test.cpp`
- `src/gsAssembler/gsVisitorNitsche.h`
- `src/gsAssembler/gsVisitorNeumannBiharmonic.h`
- `src/gsAssembler/gsVisitorNeumann.h`
- `src/gsAssembler/gsExprHelper.h`
- `src/gsAssembler/gsExprEvaluator.h`
2. Add L2projection class:
- `src/gsUtils/gsL2Projection.h`
- `src/gsUtils/gsL2Projection.hpp`
- `src/gsUtils/gsL2Projection_.cpp`
- `src/gismo.h`
3. Add direction to refinement (and degree). Affects:
- `src/gsTensor/gsTensorBasis.h`
- `src/gsTensor/gsTensorBasis.hpp`
- `src/gsNurbs/gsTensorNurbsBasis.h`
- `src/gsNurbs/gsBSplineBasis.h`
- `src/gsNurbs/gsBSplineBasis.hpp`
- `src/gsHSplines/gsHTensorBasis.h`
- `src/gsHSplines/gsHTensorBasis.hpp`
- `src/gsCore/gsRationalBasis.h`
- `src/gsHSplines/gsHBSpline_.cpp`
- `src/gsCore/gsMultiPatch.h`
- `src/gsCore/gsMultiPatch_.cpp`
- `src/gsCore/gsGeometry.h`
- `src/gsCore/gsGeometry.hpp`
- `src/gsCore/gsBasis.h`
- `src/gsCore/gsBasis.hpp`
4. Add matchwith with an offset. Affects:
- `src/gsTensor/gsTensorBasis.h`
- `src/gsTensor/gsTensorBasis.hpp` 
- `src/gsCore/gsBasis.h`
- `src/gsCore/gsBasis.hpp`
5. Add small functions to gsNurbsCreator to rotate, make grid etc
- `src/gsNurbs/gsNurbsCreator.h`
- `src/gsNurbs/gsNurbsCreator.hpp`
6. Support for gsMappedSingleXXX for plotting:
- `src/gsMSplines/gsMappedSingleBasis.h`
- `src/gsMSplines/gsMappedSingleSpline.h`
- `src/gsCore/gsPiecewiseFunction.h`
- `src/gsCore/gsFunctionSet.h`
- `src/gsCore/gsFunctionSet.hpp`
7. Plot of two `gsFunctionSet`s using Paraview
- `src/gsIO/gsWriteParaview.h`
- `src/gsIO/gsWriteParaview.hpp`
- `src/gsIO/gsWriteParaview_.cpp`
8. Fix DOF mapping for mapped splines and bases. Affects:
- `src/gsCore/gsDofMapper.h`
- `src/gsAssembler/gsExpressions.h`
9. Option in EV computation; whether to compute EV on boundaries or not
- `src/gsCore/gsBoxTopology.cpp`
- `src/gsCore/gsBoxTopology.h`
11. Access (private) data repo
- `.gitignore`
12. Typos and small fixes:
- `src/gsMSplines/gsWeightMapper.h`
- `src/gsIO/gsFileData_.cpp`
- `src/gsCore/gsBoundary.cpp`
- `src/gsAssembler/gsDirichletValues.h` (compiler issues)
- `filedata/surfaces/scordelis_lo_roof.xml`
- `extensions/gsSpectra/gsSpectra.h`

**Review by filiatra — APPROVED** [review]
